app/routes/task.py

```diff
@@ -0,0 +1,143 @@
+        completed_at=request_body['completed_at']
+        # goal_id=
+    )
+    
+    db.session.add(task_to_create)
+    db.session.commit()
+
+    response_body = {
+        'task': task_to_create.to_dict()
+    }
+
+    return make_response(jsonify(response_body), 201)
+
+@task_bp.route('/<task_id>', methods=['GET'])
+def read_single_task(task_id):
+    task = Task.query.get(task_id)
+    if not task:
+        return make_response('', 404)
+        
+    response_body = {'task': task.to_dict()} if not task.goal_id else {'task': task.to_dict_with_goal()}
+    return make_response(jsonify(response_body), 200)
+
+@task_bp.route('/<task_id>', methods=['PUT'])
+def update_single_task(task_id):
+    request_body = request.get_json()
+    task = Task.query.get(task_id)
+    if not task:
+        return make_response('', 404)
+
+    if 'title' not in request_body or 'description' not in request_body:
+        return make_response('Invalid data', 400)
+        
+    task.title = request_body['title']
+    task.description = request_body['description'] 
+
+    db.session.commit()
+
+    response_body = {
+        'task': Task.query.get(task_id).to_dict()
+    }
+    return make_response(jsonify(response_body), 200)
+
+@task_bp.route('/<task_id>/mark_complete', methods=['PATCH'])
+def update_single_task_complete(task_id):
+    # request_body = request.get_json()
+    task = Task.query.get(task_id)
+    if not task:
+        return make_response('', 404)
+        
+    task.completed_at = datetime.utcnow()
+    db.session.commit()
+
+    # "Someone just completed the task My Beautiful Task"
+    slack_channel_id = os.environ.get('SLACK_CHANNEL_ID')
+    slack_bot_token = os.environ.get('SLACK_BOT_TOKEN')
+    requests.post(
+        'https://slack.com/api/chat.postMessage',
+        headers={'Authorization': f'Bearer {slack_bot_token}'},
+        data={
+            'channel': slack_channel_id,
+            'text': f'Someone just completed the task {task.title}'
+        }
+    )
+
+    response_body = {
+        'task': task.to_dict()
+    }
+
+    return make_response(jsonify(response_body), 200)
+
+@task_bp.route('/<task_id>/mark_incomplete', methods=['PATCH'])
+def update_single_task_incomplete(task_id):
+    # request_body = request.get_json()
+    task = Task.query.get(task_id)
+    if not task:
+        return make_response('', 404)
+        
+    task.completed_at = None
+    db.session.commit()
+
+    response_body = {
+        'task': task.to_dict()
+    }
+
+    return make_response(jsonify(response_body), 200)
+
+@task_bp.route('/<task_id>', methods=['DELETE'])
+# def delete_single_task(task_id):
+#     task = Task.query.get(task_id)
+#     if not task or not taks.accessible:
+#         return make_response('', 404)
+#     task.accessible = False
+#     db.session.commit()
+#     response_body = {
+#         'details': f'Task {task_id} "{task.title}" successfully deleted'
+#     }
+#     return make_response(jsonify(response_body), 200)
```

Comment:
I like the idea here of not actually deleting tasks from the DB and instead hiding them! This is what a lot of real-world applications do. However, if you're ever doing this in a real job you have to be very careful. Different countries have different laws about data actually being deleted from the database when the customer requests it to be removed. This is especially true if you have in users in the EU because of [GDPR](https://www.cmswire.com/customer-experience/an-introduction-to-the-gdpr/).